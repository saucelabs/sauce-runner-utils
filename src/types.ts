export type Suite = {
  name: string;
  env?: { [key: string]: string };
  config?: {
    env?: { [key: string]: string };
  };
};

export type Registry = {
  scope?: string;
  url: string;
  authToken?: string;
  auth?: string;
  username?: string;
  password?: string;
  email?: string;
};

export type NpmConfig = {
  /**
   * @deprecated: registry should be avoided in favor of a "registries" entry.
   */
  registry?: string;
  registries?: Registry[];
  strictSSL?: boolean | null;
  packageLock?: boolean | string | null;
  packages?: { [key: string]: string | number };
  legacyPeerDeps?: boolean | string | null;
};

export interface PathContainer {
  path: string;
}

export interface NpmConfigContainer {
  npm?: NpmConfig;
}

export interface SuitesContainer {
  suites: Suite[];
}

export interface NodeContext {
  nodePath: string;
  npmPath: string;

  // Ignore the specific binary paths provided and rely on the binaries
  // from PATH.
  useGlobals?: boolean;
}
