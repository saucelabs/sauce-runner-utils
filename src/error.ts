export function errCode(obj: unknown): string | number {
  if (typeof obj !== 'object' || obj === null || !('code' in obj)) {
    return '';
  }
  if (typeof obj.code === 'number' || typeof obj.code === 'string') {
    return obj.code;
  }
  return '';
}
