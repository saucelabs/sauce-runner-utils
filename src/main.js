module.exports = require('./utils');
module.exports.saucectl = require('./saucectl');