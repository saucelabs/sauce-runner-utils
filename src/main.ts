/* istanbul ignore file */
// import utils from './utils';
// import saucectl from './saucectl';

// export default import from './utils';
// export saucectl { default as saucectl } from './saucectl';
