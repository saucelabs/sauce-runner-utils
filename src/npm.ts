import { spawn } from 'child_process';
import { lstat, rename, rm, writeFile } from 'fs/promises';
import { IHasNodePath } from './types';

const temporarilyMovedFiles: {[key: string]: string} = {
  'package.json': `.package.json-${process.pid}`,
  'package-lock.json': `.package-lock.json-${process.pid}`,
};

export default class NPM {
  static async removePackageJson () {
    for (const [name, replacement] of Object.entries(temporarilyMovedFiles)) {
      try {
        // Note: we don't need the return value.
        // lstat is only used to check that the file exists.
        await lstat(name);
        await rename(name, replacement);
      } catch (e) {
        console.debug(`no ${name} to archive`);
      }
    }
  }

  static async restorePackageJson () {
    for (const [name, replacement] of Object.entries(temporarilyMovedFiles)) {
      try {
        // Note: we don't need the return value.
        // lstat is only used to check that the file exists.
        await lstat(replacement);
        await rename(replacement, name);
      } catch (e) {
        console.debug(`no ${name} was archived`);
      }
    }
  }

  public static async install (nodePath: IHasNodePath, pkg: {[key: string]: string}) {
    await this.removePackageJson();
    await writeFile('package.json', JSON.stringify({
      dependencies: pkg,
    }));

    const p = spawn(nodePath.nodePath, [nodePath.npmPath, 'install']);
    p.stdout.pipe(process.stdout);
    p.stderr.pipe(process.stderr);

    const exitPromise = new Promise((resolve) => {
      p.on('exit', (exitCode) => {
        resolve(exitCode);
      });
    });

    const exitCode = await exitPromise;

    await rm('package.json', { force: true });
    await this.restorePackageJson();

    return exitCode;
  }

  public static configure (nodePath: IHasNodePath, cfg: {[key: string]: object | string | number | boolean | null }): Promise<number | null> {
    return new Promise((resolve) => {
      const args = Object.keys(cfg).map((k,) => `${k}=${cfg[k]}`);
      const p = spawn(nodePath.nodePath, [nodePath.npmPath, 'config', 'set', ...args]);
      p.stdout.pipe(process.stdout);
      p.stderr.pipe(process.stderr);
      p.on('exit', () => {
        console.log('Finished');
        resolve(0);
      });
    });
  }

  public static rebuild (nodePath: IHasNodePath, ...args: string[]): Promise<number | null> {
    return new Promise((resolve) => {
      const p = spawn(nodePath.nodePath, [nodePath.npmPath, 'rebuild', ...args]);
      p.stdout.pipe(process.stdout);
      p.stderr.pipe(process.stderr);
      p.on('exit', (exitCode) => {
        resolve(exitCode);
      });
    });
  }
}

