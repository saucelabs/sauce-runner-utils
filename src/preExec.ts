import ChildProcess from 'child_process';
import os from 'os';

function spawnAsync (cmd: string, args: string[]) {
  return new Promise(function (resolve) {
    const proc = ChildProcess.spawn(cmd, args);
    proc.stdout.on('data', (data) => {
      process.stdout.write(data.toString());
    });
    proc.stderr.on('data', (data) => {
      process.stderr.write(data.toString());
    });
    proc.on('exit', function (exitCode) {
      resolve(exitCode);
    });
    proc.on('error', function (err) {
      console.log(`Unable to start command: ${err}`);
      resolve(1);
    });
  });
}

async function preExecRunner (preExecs: string[]) {
  const cmdInvoker = (os.platform() === 'win32') ? 'cmd' : 'sh';
  const cmdArg = (os.platform() === 'win32') ? '/C' : '-c';

  for (const command of preExecs) {
    console.log(`Executing pre-exec command: ${command}`);
    const exitCode = await spawnAsync(cmdInvoker, [cmdArg, command]);
    console.log('\n');

    if (exitCode !== 0) {
      return false;
    }
  }
  return true;
}

export async function run (suite: { preExec: string[] | undefined | null }, timeoutSec: number): Promise<boolean> {
  if (!suite.preExec) {
    return true;
  }

  let timeout;
  const timeoutPromise = new Promise((resolve) => {
    timeout = setTimeout(() => {
      console.error(`Pre-Exec timed out after ${timeoutSec} seconds`);
      resolve(false);
    }, timeoutSec * 1000);
  });
  const hasPassed: boolean = await Promise.race([timeoutPromise, preExecRunner(suite.preExec)]) as boolean;
  clearTimeout(timeout);
  return hasPassed;
}

export default run;
