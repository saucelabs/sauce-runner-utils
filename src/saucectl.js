const fs = require('fs');

// Store file containing job-details url.
// Path has to match the value of the Dockerfile label com.saucelabs.job-info !
const OUTPUT_FILE_PATH = '/tmp/output.json';

function exportValue (payload) {
  fs.writeFileSync(OUTPUT_FILE_PATH, JSON.stringify(payload));
}

function updateExportedValue (data) {
  let fileData;
  try {
    const st = fs.statSync(OUTPUT_FILE_PATH);
    if (st.isFile()) {
      fileData = JSON.parse(fs.readFileSync(OUTPUT_FILE_PATH)) || {};
    }
  } catch (e) {}
  fileData = { ...fileData, ...data };
  exportValue(fileData);
}

module.exports = {
  OUTPUT_FILE_PATH, exportValue, updateExportedValue,
};