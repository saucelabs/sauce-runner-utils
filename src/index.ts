import * as saucectl from './saucectl';
import * as npm from './npm';
import * as utils from './utils';
import {
  getAbsolutePath,
  shouldRecordVideo,
  loadRunConfig,
  getDefaultRegistry,
  setUpNpmConfig,
  installNpmDependencies,
  rebuildNpmDependencies,
  hasNodeModulesFolder,
  getNpmConfig,
  prepareNpmEnv,
  getArgs,
  getEnv,
  getSuite,
  renameScreenshot,
  renameAsset,
  escapeXML
} from './utils';

export {
  npm,
  saucectl,
  utils,

  // Exporting all to keep compatibility with previous API
  getAbsolutePath,
  shouldRecordVideo,
  loadRunConfig,
  getDefaultRegistry,
  setUpNpmConfig,
  installNpmDependencies,
  rebuildNpmDependencies,
  hasNodeModulesFolder,
  getNpmConfig,
  prepareNpmEnv,
  getArgs,
  getEnv,
  getSuite,
  renameScreenshot,
  renameAsset,
  escapeXML
};
