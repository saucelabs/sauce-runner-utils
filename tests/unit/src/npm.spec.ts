import spawk from 'spawk';

import { Stats } from 'fs';

jest.mock('fs/promises');
import fs from 'fs/promises';
const fsMocked = fs as jest.Mocked<typeof fs>;

import NPM from '../../../src/npm';
import { NodeContext } from '../../../src/types';

describe('NPM', function () {
  const nodePath: NodeContext = { nodePath: 'node-bin', npmPath: 'npm-bin' };

  beforeEach(function () {
    spawk.load();
    spawk.preventUnmatched();
    jest.resetAllMocks();
  });

  afterEach(function () {
    spawk.unload();
  });

  it('.configure must invoke npm config set', async function () {
    const interceptor = spawk.spawn(nodePath.nodePath).stdout('npm runned').exit(0);
    await NPM.configure(nodePath, { registry: 'myregistry' });

    expect(interceptor.calledWith.command).toEqual(nodePath.nodePath);
    expect(interceptor.calledWith.args).toEqual([nodePath.npmPath, 'config', 'set', 'registry=myregistry']);
  });

  it('.rebuild must invoke npm rebuild', async function () {
    const interceptor = spawk.spawn(nodePath.nodePath).stdout('npm runned').exit(0);
    await NPM.rebuild(nodePath);
    expect(interceptor.calledWith.command).toEqual(nodePath.nodePath);
    expect(interceptor.calledWith.args).toEqual([nodePath.npmPath, 'rebuild']);
  });

  it('.install must invoke npm install', async function () {
    const interceptor = spawk.spawn(nodePath.nodePath).stdout('npm runned').exit(0);
    fsMocked.lstat.mockRejectedValue('non-existing');
    let writeFile, writeContent;
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    fsMocked.writeFile.mockImplementation(function (name: any, data: any): Promise<void> {
      writeFile = name;
      writeContent = data;
      return new Promise((resolve) => {
        resolve();
      });
    });
    await NPM.install(nodePath, {
      cypress: '12.6.0'
    });
    expect(interceptor.calledWith.command).toEqual(nodePath.nodePath);
    expect(interceptor.calledWith.args).toEqual([nodePath.npmPath, 'install']);
    expect(fsMocked.lstat).toBeCalledTimes(4);
    expect(fsMocked.writeFile).toBeCalledTimes(1);
    expect(writeFile).toEqual('package.json');
    expect(writeContent).toEqual(JSON.stringify({dependencies: { cypress: '12.6.0' }}));
  });

  it('.install moves package.json / package-lock.json', async function () {
    const interceptor = spawk.spawn(nodePath.nodePath).stdout('npm runned').exit(0);
    fsMocked.lstat.mockRejectedValue('non-existing');

    fsMocked.lstat.mockResolvedValue({} as Stats);
    fsMocked.rename.mockResolvedValue();
    fsMocked.writeFile.mockResolvedValue();

    await NPM.install(nodePath, {
      cypress: '12.6.0'
    });
    expect(interceptor.calledWith.command).toEqual(nodePath.nodePath);
    expect(interceptor.calledWith.args).toEqual([nodePath.npmPath, 'install']);
    expect(fsMocked.rename.mock.calls).toEqual([
      ['package.json', `.package.json-${process.pid}`],
      ['package-lock.json', `.package-lock.json-${process.pid}`],
      [`.package.json-${process.pid}`, 'package.json'],
      [`.package-lock.json-${process.pid}`, 'package-lock.json'],
    ]);
    expect(fsMocked.lstat.mock.calls).toEqual([
      ['package.json'],
      ['package-lock.json'],
      [`.package.json-${process.pid}`],
      [`.package-lock.json-${process.pid}`],
    ]);
    expect(fsMocked.writeFile.mock.calls).toEqual([
      ['package.json', JSON.stringify({dependencies: { cypress: '12.6.0' }})],
    ]);
  });
});
