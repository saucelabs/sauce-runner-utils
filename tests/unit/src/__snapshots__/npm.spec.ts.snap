// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`npm install should invoke npm 1`] = `
Array [
  Array [
    Array [
      "sl-dummy-package@2.0.0",
    ],
    [Function],
  ],
]
`;

exports[`npm rebuild should invoke npm 1`] = `
Array [
  Array [
    Array [],
    [Function],
  ],
]
`;
